theano/tensor/nnet/nnet.py

```diff
@@ -750,21 +899,53 @@ def c_code_cache_version():
 @gof.local_optimizer([tensor.Elemwise])
 def local_logsoftmax(node):
     """
-    Detect Log(Softmax(x)) and replace it with LogSoftmax(x)
-
+    Detect Log(Softmax(x)) and replace it with LogSoftmax(x) and
+    detect Log(Softmax(x)[*idx]) and replace it with LogSoftmax(x)[*idx]
     Note: only forward pass is affected
     """
+    list_classes = ((Subtensor, AdvancedSubtensor1, AdvancedSubtensor))
     if (isinstance(node.op, tensor.Elemwise) and
             isinstance(node.op.scalar_op, scalar.basic.Log) and
-            len(node.inputs) == 1 and
-            node.inputs[0].owner is not None and
-            isinstance(node.inputs[0].owner.op, Softmax)):
-        inVars = node.inputs[0].owner.inputs[0]
-        new_op = LogSoftmax()
-        ret = new_op(inVars)
-        ret .tag.values_eq_approx = values_eq_approx_remove_inf
-        copy_stack_trace([node.inputs[0], node.outputs[0]], ret)
-        return [ret]
+            node.inputs[0].owner is not None):
+
+            # Case log(softmax(x))
+            if isinstance(node.inputs[0].owner.op, Softmax):
+                softmax_input = node.inputs[0].owner.inputs[0]
+                axis = node.inputs[0].owner.op.axis
+                ret = LogSoftmax(axis)(softmax_input)
+                ret.tag.values_eq_approx = values_eq_approx_remove_inf
+                copy_stack_trace([node.inputs[0], node.outputs[0]], ret)
+                return [ret]
+
+            if isinstance(node.inputs[0].owner.op, tensor.elemwise.DimShuffle) and isinstance(node.inputs[0].owner.inputs[0].owner.op, list_classes):
+                node = node.inputs[0].owner
```

Comment:
I'm wondering, aren't you completely removing the DimShuffle node here and on line 929?

Reply:
I did this because those dimshuffle are canceling each other. They happens in order to be managed by the subtensor op which can be used only on a specific dimension. It looks like dimshuffle(subtensor(dimshuffle(x)). So I though that, since I am creating a new subtensor, it will apply back the dimshuffle himself.